### ISSUE TITLE
ROI tool: Wishlist

### ISSUE THREAD

**Issue by Lestropie** [jdtournier/mrtrix3#155]
In no particular order. Feel free to append.

1. 3D brush: Requires 3D data for undo/redo operations. I wonder if storing diffs as lists of voxels would be more efficient than a 3D texture for each? 3D brush would also enable full ROI editing when off-axis.

2. Copy to/from slice above/below buttons

3. Fill tool. This could either be clicking in the middle of an already-drawn closed loop, or an alternative drawing tool that tracks the mouse path from click to release and then fills the relevant region.

4. Label editing. If the input file is an integer image rather than boolean, allow to select the integer value  to be written by the brush. Would likely require deeper modifications in terms of texture bit depths, shaders etc..

5. Some mechanism to allow the user to indicate that they wish for the ROI to be saved as a UInt8 rather than bitwise; provide FSL compatibility without having to run mrconvert explicitly.

**Comment by jdtournier** [comment]
OK, my 2 cents:

1. I don't think there's any need for the 3D brush. Maybe someone can convince me otherwise, but in my experience it's difficult to use properly.

2. Would be good, and should be easy to implement. 

3. I was planning on including a freehand drawing tool on there eventually. Draw a loop and it gets filled when you're done. So this one is definitely planned. Incidentally, this feature is probably the biggest driver  for the current slice-wise UI design: this simply cannot be done off-axis...

4. OK, now this is starting to sound like one of the approaches I'd suggested before: a full image editor. If we're going to start doing that, I would have preferred we had a tool to edit the main image directly, and use the overlay tool (potentially in underlay mode) to select what background to guide the drawing. I'm still of the opinion that design-wise, this is a much saner approach than what we currently have, and would remove most of the weird behaviour we're having to get around at the moment...

5. Not an issue for FSL. The default behaviour currently is to create bitwise NIfTIs as uint8 anyway (can be overridden with a config file option).

**Comment by Lestropie** [comment]
1. Yeah let's see what happens after merge.

2. Agreed. Does it make more sense to copy _to_ or _from_ the current slice? Or should there be buttons for both?

3. I can have a go at this. Though I disagree with it not being possible off-axis; I implemented a process in the tractogram segmentation code that was similar, but that one needed a variable surface normal.

4. True enough. Will leave this as a separate wishlist item. Another advantage doing it that way is the potential use of colour tables for integer data.

5. Forgot about that. Cool.

**Comment by Lestropie** [comment]
2 and 3 done. Feel free to improve on the icons for slice copying, don't like mine. Fill I did as a single button click (i.e. surround the region using the brush tool, change to fill tool, click in the middle). Circling the region using a lasso-type tool then having it fill on button release may be preferable, but this way was trivial to implement.

**Comment by Lestropie** [comment]
6. Expand ray-tracing concept used when brush size is minimum (used to prevent gaps in lines drawn) to apply to brushes of any size. Would likely be implemented by first drawing the spheres at either end point, using these to determine a 3D processing window; then from each voxel centre, highlight those for which the projection to the drawn line is inside the line and within the specified radius.

**Comment by Lestropie** [comment]
6 was fixed in #203. So I think that's everything so far either ticked off or flung to the not-to-do list.

7. Slice-wise and 3D filter operations, e.g. dilate, erode, lcc, median. Would require discarding the undo stack in the case where a 3D filter is applied. (User request)

**Comment by jdtournier** [comment]
Yes, that'll be a tough one undo-wise... I've tried to limit the amount of RAM required to store the undo stack by limiting all operations to current slice only, but would clearly break that. It is a sensible request though, maybe we need to think about extending the undo stack to allow storing the entire volume for such cases...?

**Comment by Lestropie** [comment]
Well, as soon as you generalize the undo stack to 3D, the question of 3D brushes will inevitably rise again; it'd be strange to allow one but not the other. In the discussion I had, they were content to have the undo stack erased upon applying any 3D operation. If we go one step at a time, it'd make sense to do this with the undo stack erase, then later think about whether to extend the undo stack.

**Comment by jdtournier** [comment]
Well, my reluctance to allow the 3D brush is more that it's very difficult to use effectively, than the difficulty in supporting it in the undo stack (although it is an issue). I don't remember this previous discussion about erasing the undo stack, but given how mushy my brain is these days, I'm more than happy to concede that I might have said that... Nonetheless, I think the long-term plan has to be that if we have an undo stack, it works regardless of the operation. But in the interest of supporting this new feature without delaying things needlessly, happy to have it ditch the undo stack for now. We should leave this issue open until the undo stack is properly fixed though...

As to implementation, having had a quick look at the UndoEntry class, it should be fairly simple to support a full 3D undo - although it'll clearly take some work.

**Comment by Lestropie** [comment]
>  I don't remember this previous discussion about erasing the undo stack, but given how mushy my brain is these days, I'm more than happy to concede that I might have said that...

No no, this was a discussion with a software user. Though that doesn't negate the overwhelming evidence for your brain being mush. :-P  All I can say is that for this one user, they were content to lose their undo stack if it meant supporting 3D filter operations within the GUI.

**Comment by jdtournier** [comment]
How reassuring...
